### DISCUSSION TITLE
How To: Query for all values of a given Key within an array

### DISCUSSION THREAD

**Discussion by RobertLbebber** [danielaparker/jsoncons#329] [Q&A] [answered]
Given:
```
{
    "arr":[
        {
            "keyOfInterest":true,
            "keyOfInsignificants":true,
        },
        {
            "keyOfInterest":false,
            "keyOfInsignificants":true,
        },
        {
            "keyOfInterest":true,
            "keyOfInsignificants":true,
        }
    ]
}
```

How could you use json_query (or some other querying-like function) to fetch the `keyOfInterest` values? 
Ideally in the format of `[true, false, true]` or `{"0":true, "1":false, "2":true}`

**Comment by danielaparker** [comment]
With JSONPath, like this:
```c++
#include <jsoncons/json.hpp>
#include <jsoncons_ext/jsonpath/jsonpath.hpp>

using jsoncons::json;
namespace jsonpath = jsoncons::jsonpath;

int main()
{
    json root = json::parse(R"(
{
    "arr":[
        {
            "keyOfInterest":true,
            "keyOfInsignificants":true
        },
        {
            "keyOfInterest":false,
            "keyOfInsignificants":true
        },
        {
            "keyOfInterest":true,
            "keyOfInsignificants":true
        }
    ]
}
)");
    auto results = jsonpath::json_query(root, "$.arr[*]['keyOfInterest']");

    std::cout << results << "\n";
}
```
Output:
```json
[true,false,true]
```

You can think of a JSONPath expression in terms of its left hand side and its right hand side, its right hand side is applied  to each element produced by its left hand side. So in the expression
```
$.arr[*]['keyOfInterest']
```
`['keyOfInterest']` is applied to each element produced by the wildcard selector on `arr` to its left, which is all the elements of `arr`. 

Note that each element in the result - `true`, `false`, `true` - correspond to an element at a specific location in the original JSON document. This is a feature of JSONPath. You can't produce 
```json
{"0":true, "1":false, "2":true}
```
with a JSONPath expression, you can only select a collection of elements that are already in the JSON document.

**Comment by danielaparker** [comment, marked as answer]
Or using JMESPath:
```c++
#include <jsoncons/json.hpp>
#include <jsoncons_ext/jmespath/jmespath.hpp>

using jsoncons::json;
namespace jmespath = jsoncons::jmespath;

int main()
{
    json root = json::parse(R"(
{
    "arr":[
        {
            "keyOfInterest":true,
            "keyOfInsignificants":true
        },
        {
            "keyOfInterest":false,
            "keyOfInsignificants":true
        },
        {
            "keyOfInterest":true,
            "keyOfInsignificants":true
        }
    ]
}
)");

    try
    {
        auto result1 = jmespath::search(root, "arr[*].\"keyOfInterest\"");
        std::cout << "(1) " << result1 << "\n";
        auto result2 = jmespath::search(root, "arr[*].{\"Key of Interest\" : \"keyOfInterest\"}");
        std::cout << "(2) " << result2 << "\n";
    }
    catch (const std::exception& e)
    {
        std::cout << e.what() << std::endl;
    }
}
```
Output:
```json
(1) [true,false,true]
(2) [{"Key of Interest":true},{"Key of Interest":false},{"Key of Interest":true}]
```
Note that, unlike JSONPath, JMESPath has some capabilities for producing elements that are _not_ in the original JSON document.

**Reply by RobertLbebber** [reply]
Thanks, using the `jmespath::search(root, "arr[*].{\"Key of Interest\" : \"keyOfInterest\"}");` I can make object maps that carry multiple fields for a single element among a list. Like so:

`jmespath::search(root, "arr[*].{\"Key of Interest\" : \"keyOfInterest\", \"otherUnmetionedKey\": \"otherUnmetionedKeyValue\"}");`
Which yields objects of this nature:
```
[
    {
        "Key of Interest": <boolean>,
        "otherUnmetionedKey": <whatever>
    },
    // Other Elements...
]
```
